FlatLambdaCDM hypergeometric function solution, de Sitter, and Einstein-de Sitter for no radiation case

wmwv opened the pull request:
For a Flat LambdaCDM cosmology with no radiation (and no relativistic species), the comoving distance is expressible as an incomplete elliptic integral.  This pull request implements this to make a ~20x speedup in calculating distances.

This pull request makes FlatLambdaCDM set the comoving distance calculation to use the incomplete elliptic integral solution in the __init__ for the case of `Tcmb0=0`.  The speed-up is a factor of ~20x.  It depends on the number of redshifts being calculated simultaneously.  In more detail, my testing found that it was ~(1.2, 20, 100) times faster for n=(1, 200, 2000) redshifts

No new tests are required.  The existing ones already test both the `Tcmb0=0` and `Tcmb!=0` cases and this present change still passes those test.

aconley commented on astropy/cosmology/core.py:
```diff
@@ -1752,6 +1753,64 @@ def __init__(self, H0, Om0, Tcmb0=0, Neff=3.04,
+        """
+
+        from scipy.special import ellipkinc
+        F = ellipkinc
```
Why alias to F?

aconley commented on astropy/cosmology/core.py:
```diff
@@ -1752,6 +1753,64 @@ def __init__(self, H0, Om0, Tcmb0=0, Neff=3.04,
+        from scipy.special import ellipkinc
+        F = ellipkinc
+        # math.sqrt is several times faster than np.sqrt for scalars
+        phi = np.arccos((1 + (1-sqrt(3))*x) /
```
missing spaces around - and *

aconley commented on astropy/cosmology/core.py:
```diff
@@ -1752,6 +1753,64 @@ def __init__(self, H0, Om0, Tcmb0=0, Neff=3.04,
+        F = ellipkinc
+        # math.sqrt is several times faster than np.sqrt for scalars
+        phi = np.arccos((1 + (1-sqrt(3))*x) /
+                        (1 + (1+sqrt(3))*x))
```
Missing spaces around + and *

aconley commented on astropy/cosmology/core.py:
```diff
@@ -1752,6 +1753,64 @@ def __init__(self, H0, Om0, Tcmb0=0, Neff=3.04,
+        # math.sqrt is several times faster than np.sqrt for scalars
+        phi = np.arccos((1 + (1-sqrt(3))*x) /
+                        (1 + (1+sqrt(3))*x))
+        k = np.cos(np.pi/12)
```
Missing spaces around /, but really it's a constant.

aconley commented on astropy/cosmology/core.py:
```diff
@@ -1752,6 +1753,64 @@ def __init__(self, H0, Om0, Tcmb0=0, Neff=3.04,
+        phi = np.arccos((1 + (1-sqrt(3))*x) /
+                        (1 + (1+sqrt(3))*x))
+        k = np.cos(np.pi/12)
+        m = k*k  # np.cos(np.pi/12)**2 == 1/2 + math.sqrt(3)/4
```
Please just hardwire the value with the comment rather than recomputing each time.  And in that case you can omit the packages from the comment:

m = 0.9330127018922193  # cos(pi/12)**2

aconley commented on astropy/cosmology/core.py:
```diff
@@ -1752,6 +1753,64 @@ def __init__(self, H0, Om0, Tcmb0=0, Neff=3.04,
+        k = np.cos(np.pi/12)
+        m = k*k  # np.cos(np.pi/12)**2 == 1/2 + math.sqrt(3)/4
+        # ellipkinc expects m=k*k as its second argument.
+        return 3**(-1./4) * F(phi, m)
```
Just use ellipkinc:
 return 3 ** (-0.25) * ellipkinc(phi, m)

aconley commented on astropy/cosmology/core.py:
```diff
@@ -1752,6 +1753,64 @@ def __init__(self, H0, Om0, Tcmb0=0, Neff=3.04,
+                        (1 + (1+sqrt(3))*x))
+        k = np.cos(np.pi/12)
+        m = k*k  # np.cos(np.pi/12)**2 == 1/2 + math.sqrt(3)/4
+        # ellipkinc expects m=k*k as its second argument.
```
Comment unnecessary.

aconley commented on astropy/cosmology/core.py:
```diff
@@ -1752,6 +1753,64 @@ def __init__(self, H0, Om0, Tcmb0=0, Neff=3.04,
+
+        s = ((1-self._Om0)/self._Om0) ** (1/3)
+        prefactor = self._hubble_distance / sqrt(s*self._Om0)
+        return prefactor * (self._T_legendre(s/(1+z1)) - self._T_legendre(s/(1+z2)))
```
Missing spaces around operators.

aconley commented on astropy/cosmology/core.py:
```diff
@@ -1752,6 +1753,64 @@ def __init__(self, H0, Om0, Tcmb0=0, Neff=3.04,
+                raise ValueError(msg)
+
+        s = ((1-self._Om0)/self._Om0) ** (1/3)
+        prefactor = self._hubble_distance / sqrt(s*self._Om0)
```
spaces around *

bsipocz commented on astropy/cosmology/core.py:
```diff
@@ -1752,6 +1753,64 @@ def __init__(self, H0, Om0, Tcmb0=0, Neff=3.04,
+                msg = "In _comoving_distance_z1z2: z1 and z2 have different lengths"
+                raise ValueError(msg)
+
+        s = ((1-self._Om0)/self._Om0) ** (1/3)
```
This is fine as this PR can go into v3.1 the earliest, and we've already dropped python <3.5 in v3.0.

wmwv commented:
@aconley Thanks for the review.

I realized that the elliptic integral formulation fails for Om0=0 (de Sitter) and Om0=1 (Einstein-de Sitter) cosmologies.  I added a test to `tests/test_cosmology.py` to check this, and added functions to implement these special cases and expanded the check in `FlatLambdaCDM.__init__` to assign the appropriate function for these cases.

aconley commented:
There also seems to be a problem whenever Omega_m > 1, since your s factor becomes negative.  I'm not sure if there's a good way to save that, or if you should restrict your solution purely to the Om (0,1) interval.

wmwv commented:
Ah, yes.  Good point.  

* Test added for FlatLambdaCDM, Omega_M>1 to make sure distances still work.
* Added a hypergeometric form that is valid for all Omega_M>0.  Kept Einstein - de Sitter analytic for Omega_M=1 for performance.

wmwv commented:
The hypergeometric one is is actually faster than the elliptic integral version, as tested on a set of evenly-spaced redshifts from 0.001 to 100.

The special de Sitter and Einstein-de Sitter cases are several times faster than the hypergeometric calculation.  This is expected, although the fact that it's only a factor of a few is almost surprising.

```
INTEGRAL
Time to calculate comoving distance for 2000 redshifts.
                       General : 0.0726 s
          Einstein - de Sitter : 0.0764 s
                     de Sitter : 0.0212 s
                   Omega_M > 1 : 0.0902 s
HYPERGEOMETRIC
Time to calculate comoving distance for 2000 redshifts.
                       General : 0.00044 s
          Einstein - de Sitter : 0.00016 s
                     de Sitter : 6.33e-05 s
                   Omega_M > 1 : 0.0005 s
ELLIPTIC
Time to calculate comoving distance for 2000 redshifts.
                       General : 0.000708 s
          Einstein - de Sitter : 0.000162 s
                     de Sitter : 5.98e-05 s
TypeError: ufunc 'ellipkinc' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
                   Omega_M > 1 : nan s
```

wmwv commented:
I removed the _elliptic and _T_legendre functions and kept only the hypergeometric code.

pllim commented:
Thank you, all! Merging now.
